pallets/farming/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+        #[pallet::weight(0)]
+        pub fn force_retire_pool(origin: OriginFor<T>, pid: PoolId) -> DispatchResult {
+            T::ControlOrigin::ensure_origin(origin)?;
+
+            let mut pool_info = Self::pool_infos(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+            ensure!(
+                PoolState::state_valid(Action::ForceRetirePool, pool_info.state),
+                Error::<T>::InvalidPoolState
+            );
+
+            let withdraw_limit_time = BlockNumberFor::<T>::default();
+            let retire_limit = RetireLimit::<T>::get();
+            let mut all_retired = true;
+            let share_infos = SharesAndWithdrawnRewards::<T>::iter_prefix_values(pid);
+            for (retire_count, share_info) in share_infos.enumerate() {
+                if retire_count.saturated_into::<u32>() >= retire_limit {
+                    all_retired = false;
+                    break;
+                }
+                let who = share_info.who;
+                Self::remove_share(&who, pid, None, withdraw_limit_time)?;
+                Self::claim_rewards(&who, pid)?;
+                if let Some(ref gid) = pool_info.gauge {
+                    Self::gauge_claim_inner(&who, *gid)?;
+                }
+                Self::process_withdraw_list(&who, pid, &pool_info, true)?;
+            }
+
+            if all_retired {
+                if let Some(ref gid) = pool_info.gauge {
+                    let mut gauge_pool_info =
+                        Self::gauge_pool_infos(gid).ok_or(Error::<T>::GaugePoolNotExist)?;
+                    gauge_pool_info.gauge_state = GaugeState::Unbond;
+                    GaugePoolInfos::<T>::insert(gid, gauge_pool_info);
+                }
+                pool_info.state = PoolState::Retired;
+                pool_info.gauge = None;
+                PoolInfos::<T>::insert(pid, pool_info);
+                Self::deposit_event(Event::AllRetired { pid });
+            } else {
+                Self::deposit_event(Event::PartiallyRetired { pid });
+            }
+            Ok(())
+        }
+
+        #[pallet::call_index(7)]
+        #[pallet::weight(0)]
+        pub fn set_retire_limit(origin: OriginFor<T>, limit: u32) -> DispatchResult {
+            T::ControlOrigin::ensure_origin(origin)?;
+
+            RetireLimit::<T>::mutate(|old_limit| {
+                *old_limit = limit;
+            });
+
+            Self::deposit_event(Event::RetireLimitSet { limit });
+            Ok(())
+        }
+
+        #[pallet::call_index(8)]
+        #[pallet::weight(0)]
+        pub fn close_pool(origin: OriginFor<T>, pid: PoolId) -> DispatchResult {
+            T::ControlOrigin::ensure_origin(origin)?;
+
+            let mut pool_info = Self::pool_infos(pid).ok_or(Error::<T>::PoolDoesNotExist)?;
+            ensure!(
+                PoolState::state_valid(Action::ClosePool, pool_info.state),
+                Error::<T>::InvalidPoolState
+            );
+
+            pool_info.state = PoolState::Dead;
```

Comment:
What does dead mean? Does it mean no one use it any more?

Reply:
Suppose we don't want to incentive a farming pool, we need first close the pool, then we can retire and kill the pool.
In Dead state, user can still claim rewards and withdraw tokens that staked into farming. but can't deposit token anymore.